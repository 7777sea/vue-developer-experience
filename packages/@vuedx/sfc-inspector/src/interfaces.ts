import { NodePath } from '@babel/traverse'
import {
  CallExpression,
  ExportDefaultDeclaration,
  File,
  ObjectExpression,
} from '@babel/types'
import { RootNode } from '@vue/compiler-core'
import {
  SFCScriptBlock,
  SFCTemplateBlock,
  SFCDescriptor,
} from '@vue/compiler-sfc'
import { EventInfo, PropInfo } from './VueComponentInfo'

export interface EnhancedSFCScriptBlock extends SFCScriptBlock {
  ast: File
  paths: ComponentOptionsPaths
}

export interface EnhancedSFCTemplateBlock extends SFCTemplateBlock {
  ast: RootNode
}

export interface EnhancedSFCDescriptor {
  filename: string
  script: EnhancedSFCScriptBlock | null
  template: EnhancedSFCTemplateBlock | null
  styles: SFCDescriptor['styles']
  customBlocks: SFCDescriptor['customBlocks']
}

export interface SourceFile {
  content: string
  filename: string
}
export interface InspectorOptions {
  onError(error: Error): void
  resolve(imported: string, importee: string): Promise<string>
  load(id: string): Promise<SourceFile>
  plugins: InspectorPlugin[]
}
export interface InspectorPlugin {
  (
    descriptor: EnhancedSFCDescriptor,
    context: InspectorContext
  ): void | Promise<void>
}
export interface InspectorContext {
  data: any
  addError(error: Error): void
  addProp(prop: PropInfo): void
  addEvent(event: EventInfo): void
}

export type ComponentOptionsPaths =
  | {
      default: NodePath<ExportDefaultDeclaration>
      options: NodePath<ObjectExpression>
      define: NodePath<CallExpression>
    }
  | {
      default: NodePath<ExportDefaultDeclaration>
      options: NodePath<ObjectExpression>
      define: null
    }
  | {
      default: NodePath<ExportDefaultDeclaration> | null
      options: null
      define: null
    }
