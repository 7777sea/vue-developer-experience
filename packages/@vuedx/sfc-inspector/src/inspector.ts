import { parse } from './parser'

export async function baseInspect(
  source: SourceFile,
  options: InspectorOptions
) {
  const descriptor = parse({
    source: source.content,
    sourceFilename: source.filename,
    onError(error) {
      options.onError(error)
    },
  })

  const info: VueComponentInfo = {
    props: [],
    events: [],
  }

  const context: InspectorContext = {
    data: info,
    addError(error: Error) {
      options.onError(error)
    },
    addProp(prop: PropInfo) {
      const index = info.props.findIndex(cur => cur.name === prop.name)

      if (index >= 0) {
        info.props.splice(index, 1, mergeProp(info.props[index], prop))
      } else {
        info.props.push(prop)
      }
    },
    addEvent(event: EventInfo) {
      const index = info.events.findIndex(cur => cur.name === event.name)

      if (index >= 0) {
        info.events.splice(index, 1, mergeEvent(info.events[index], event))
      } else {
        info.events.push(event)
      }
    },
  }

  await Promise.all(options.plugins.map(plugin => plugin(descriptor, context)))

  return { descriptor, info }
}

function mergeProp(target: PropInfo, source: PropInfo) {
  if (source.default) target.default = source.default
  if (source.required != undefined) target.required = source.required
  if (source.description) target.description = source.description
  if (source.type) target.type = { ...target.type, ...source.type }

  return target
}
function mergeEvent(target: EventInfo, source: EventInfo) {
  if (source.description) target.description = source.description
  if (source.type) {
    // @ts-ignore - hidden property.
    if (!target.type.previous) {
      Object.defineProperty(target.type, 'previous', {
        enumerable: false,
        value: target.type.static === 'any' ? [] : [target.type.static],
      })
    }
    // @ts-ignore - hidden property.
    const types: string[] = target.type.previous

    if (!types.includes(source.type.static) && source.type.static !== 'any') {
      types.push(source.type.static)

      target.type.static = types.join('|')
    }
  }

  return target
}

import { SourceFile, InspectorOptions, InspectorContext } from './interfaces'
import { VueComponentInfo, PropInfo, EventInfo } from './VueComponentInfo'
