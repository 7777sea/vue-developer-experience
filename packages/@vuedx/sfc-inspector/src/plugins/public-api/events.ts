import traverse, { NodePath } from '@babel/traverse'
import { Comment, TSAsExpression, CallExpression } from '@babel/types'
import { findObjectMethod, stringifyBabelAST } from '../../ast/babel'
import { parseJSDoc, stringifyJSDocAST } from '../../ast/jsdoc'
import {
  ComponentOptionsPaths,
  EnhancedSFCScriptBlock,
  InspectorContext,
} from '../../interfaces'
import { EventInfo } from '../../VueComponentInfo'
import { BlockStatement } from '@vue/compiler-core'

export function inspectEvents(
  script: EnhancedSFCScriptBlock,
  context: InspectorContext
) {
  inspectAnnotationEmits(script.paths, context)
  inspectDollorEmits(script.paths, context)
  inspectSetupEmits(script.paths, context)
}

function inspectDollorEmits(
  paths: ComponentOptionsPaths,
  context: InspectorContext
) {
  if (paths.options) {
    traverse(
      paths.options.node as any,
      {
        CallExpression(node$) {
          if (node$.isCallExpression()) {
            const callee$ = node$.get('callee')

            if (callee$.isMemberExpression()) {
              const object$ = callee$.get('object')
              const property$ = callee$.get('property')

              if (
                object$.isThisExpression() &&
                !Array.isArray(property$) &&
                property$.isIdentifier({ name: '$emit' })
              ) {
                const args = node$.get('arguments')

                inspectCallExpressionArguments(node$, args, context)
              }
            }
          }
        },
      },
      paths.options.scope,
      paths.options.parentPath
    )
  }
}

function inspectAnnotationEmits(
  paths: ComponentOptionsPaths,
  context: InspectorContext
) {
  if (!paths.default) return
  const comments: Comment[] = []

  if (paths.default.node.leadingComments) {
    comments.push(...paths.default.node.leadingComments)
  }

  if (paths.define?.node.leadingComments) {
    comments.push(...paths.define.node.leadingComments)
  }

  if (paths.options?.node.leadingComments) {
    comments.push(...paths.options.node.leadingComments)
  }

  if (comments.length) {
    const ast = parseJSDoc(prepareDocComment(comments))

    ast.tags.forEach((tag) => {
      if (tag.title === 'emits' && tag.name) {
        context.addEvent({
          name: tag.name,
          description: tag.description || '',
          type: {
            static: tag.type ? stringifyJSDocAST(tag.type) : 'any',
          },
        })
      }
    })
  }
}

function inspectSetupEmits(
  paths: EnhancedSFCScriptBlock['paths'],
  context: InspectorContext
) {
  if (!paths.options) return

  const setup$ = findObjectMethod(paths.options, 'setup')

  if (setup$) {
    const params = setup$.get('params') as NodePath<any>[]

    if (params.length >= 2) {
      const context$ = params[1]

      const isEmitExpression = createEmitExpressionChecker(context$)

      traverse(
        setup$.node.body as any,
        {
          CallExpression(node$) {
            if (isEmitExpression(node$)) {
              inspectCallExpressionArguments(
                node$,
                node$.get('arguments'),
                context
              )
            }
          },
        },
        (setup$.get('body') as NodePath<BlockStatement>).scope,
        setup$
      )
    }
  }
}

function createEmitExpressionChecker(context$: NodePath<any>) {
  if (context$.isObjectPattern()) {
    const emit$ = context$.get('properties').find((property$) => {
      if (property$.isObjectProperty()) {
        const key$ = property$.get('key')

        if (!Array.isArray(key$) && key$.isIdentifier({ name: 'emit' })) {
          return true
        }
      }
      return false
    })

    if (emit$?.isObjectProperty()) {
      const value$ = emit$.get('value')

      if (value$.isIdentifier()) {
        const name = value$.node.name

        return (node$: NodePath<any>) => {
          return (
            node$.isCallExpression() &&
            node$.get('callee').isIdentifier({ name })
          )
        }
      }
    }

    const rest$ = context$
      .get('properties')
      .find((property$) => property$.isRestElement())

    if (rest$?.isRestElement()) context$ = rest$.get('argument')
  }

  if (context$.isIdentifier()) {
    const name = context$.node.name // context

    return (node$: NodePath<any>) => {
      if (node$.isCallExpression()) {
        const callee$ = node$.get('callee')
        if (callee$.isMemberExpression()) {
          const object$ = callee$.get('object')
          const property$ = callee$.get('property')

          return (
            object$.isIdentifier({ name }) &&
            !Array.isArray(property$) &&
            property$.isIdentifier({ name: 'emit' })
          )
        }
      }

      return false
    }
  }

  return (node$: NodePath<any>) => false
}

function inspectCallExpressionArguments(
  node$: NodePath<CallExpression>,
  args: NodePath<any>[],
  context: any
) {
  const event: EventInfo = {
    name: '',
    description: '',
    type: {
      static: 'any',
    },
  }

  if (args.length > 0) {
    const eventName$ = args[0]

    if (eventName$.isStringLiteral()) {
      event.name = eventName$.node.value
    }

    if (args.length > 1) {
      const payload$ = args[1]

      if (payload$.node.leadingComments?.length) {
        const { tags } = parseJSDoc(
          prepareDocComment(payload$.node.leadingComments)
        )

        tags.forEach((tag) => {
          if (tag.type) {
            event.type = {
              static: stringifyJSDocAST(tag.type),
            }
          }
        })
      }

      if (payload$.isNodeType('TSAsExpression')) {
        event.type = {
          static: stringifyBabelAST(
            (payload$.node as TSAsExpression).typeAnnotation
          ),
        }
      }

      if (payload$.isIdentifier()) {
        // Maybe use typescript to detect type.
      }
    } else {
      event.type = {
        static: 'never',
      }
    }

    const statement$ = node$.findParent((node$) => node$.isStatement())

    if (statement$.node.leadingComments?.length) {
      event.description =
        parseJSDoc(prepareDocComment(statement$.node.leadingComments))
          .description || ''
    }
  }

  if (event.name) context.addEvent(event)
}

function prepareDocComment(comments: readonly Comment[]): string {
  return `/*${comments.map((comment) => comment.value).join('\n')}*/`
}
