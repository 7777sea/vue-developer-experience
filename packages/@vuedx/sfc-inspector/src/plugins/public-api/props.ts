import { NodePath } from '@babel/traverse'
import {
  functionExpression,
  TSTypeAnnotation,
  TSTypeLiteral,
  TSTypeParameterInstantiation,
  ArrayExpression,
  ObjectExpression,
} from '@babel/types'
import {
  findObjectProperty,
  getObjectMemberName,
  getTypeName,
  stringifyBabelAST,
} from '../../ast/babel'
import { parseJSDoc, stringifyJSDocAST } from '../../ast/jsdoc'
import { EnhancedSFCScriptBlock, InspectorContext } from '../../interfaces'
import { PropInfo } from '../../VueComponentInfo'

export function inspectProps(
  { paths }: EnhancedSFCScriptBlock,
  context: InspectorContext
) {
  inspectPropsOption(paths, context)
  inspectDefineComponentGenericProps(paths, context)
}

function inspectPropsOption(
  paths: EnhancedSFCScriptBlock['paths'],
  context: InspectorContext
) {
  if (paths.options) {
    const props$ = findObjectProperty(paths.options, 'props')?.get(
      'value'
    ) as unknown as NodePath<any>

    if (props$?.isArrayExpression()) {
      inspectArrayProps(props$, context)
    } else if (props$?.isObjectExpression()) {
      inspectObjectProps(props$, context)
    }
  }
}

function inspectArrayProps(
  props$: NodePath<ArrayExpression>,
  context: InspectorContext
) {
  ;(props$.get('elements') as NodePath<any>[]).forEach((element$) => {
    if (element$.isStringLiteral()) {
      context.addProp({
        name: element$.node.value,
        description: '',
        required: false,
        default: null,
        type: {
          runtime: 'any',
          static: 'any',
        },
        meta: {},
      })
    }
  })
}

function inspectObjectProps(
  props$: NodePath<ObjectExpression>,
  context: InspectorContext
) {
  ;(props$.get('properties') as NodePath<any>[]).forEach((property$) => {
    if (property$.isObjectProperty()) {
      const prop$ = property$.get('value')
      const prop: PropInfo = {
        name: getObjectMemberName(property$) as string,
        description: '',
        required: false,
        default: null,
        type: {
          runtime: 'any',
          static: 'any',
        },
        meta: {},
      }

      Object.assign(prop, findPropInfo(prop$))

      const comments = (property$.node.leadingComments || [])
        .filter((comment) => comment.value.startsWith('*'))
        .map((comment) => comment.value)

      if (comments.length) {
        const ast = parseJSDoc(`/*${comments.join('\n')}*/`)

        prop.description = ast.description

        ast.tags.forEach((tag) => {
          if (tag.type) {
            if (tag.title === 'type') {
              prop.type.static = stringifyJSDocAST(tag.type)
            }
          } else {
            prop.meta[tag.title] =
              tag.description == null ? true : tag.description
          }
        })
      }

      context.addProp(prop)
    }
  })
}

function inspectDefineComponentGenericProps(
  paths: EnhancedSFCScriptBlock['paths'],
  context: InspectorContext
) {
  if (paths.define?.isCallExpression()) {
    const typeParams$ = paths.define.get('typeParameters')

    if (typeParams$.isNodeType('TSTypeParameterInstantiation')) {
      const params = (typeParams$ as NodePath<
        TSTypeParameterInstantiation
      >).get('params') as NodePath<any>[]

      if (Array.isArray(params) && params.length) {
        const param$ = params[0]

        if (param$.isNodeType('TSTypeLiteral')) {
          const type$ = param$ as NodePath<TSTypeLiteral>

          ;(type$.get('members') as NodePath<any>[]).forEach((member$) => {
            const annotation$ = member$.get('typeAnnotation') as NodePath<
              TSTypeAnnotation
            >

            const key$ = member$.get('key') as NodePath<any>

            if (key$.isIdentifier()) {
              context.addProp({
                name: key$.node.name,
                type: {
                  static: stringifyBabelAST(annotation$.node.typeAnnotation),
                },
                required: !key$.node.optional,
              } as any)
            }
          })
        }
      }
    }
  }
}

function findPropInfo(node$: NodePath<any>) {
  const prop: Partial<PropInfo> = {}

  if (node$.isIdentifier()) {
    const type = getTypeName(node$.node.name)
    prop.type = {
      static: type,
      runtime: type,
    }
  } else if (node$.isArrayExpression()) {
    const type = node$
      .get('elements')
      .map((element$) =>
        element$.isIdentifier()
          ? getTypeName(element$.node.name)
          : element$.isNullLiteral()
          ? 'null'
          : stringifyBabelAST(element$.node!)
      )
      .filter(Boolean)
      .join('|')

    prop.type = {
      runtime: type,
      static: type,
    }
  } else if (node$.isObjectExpression()) {
    const type$ = findObjectProperty(node$, 'type')
    if (type$) {
      prop.type = findPropInfo(type$.get('value')).type
    }

    const required$ = findObjectProperty(node$, 'required')
    if (required$) {
      const value$ = required$.get('value') as NodePath<any>
      if (value$.isBooleanLiteral()) {
        prop.required = value$.node.value
      }
    }

    const default$ = findObjectProperty(node$, 'default')
    if (default$) {
      const value$ = default$.get('value') as NodePath<any>

      prop.default =
        value$.isFunctionExpression() || value$.isArrowFunctionExpression()
          ? {
              factory: stringifyBabelAST(value$.node),
            }
          : default$.isObjectMethod()
          ? {
              factory: stringifyBabelAST(
                functionExpression(
                  null,
                  default$.node.params,
                  default$.node.body
                )
              ),
            }
          : {
              value: stringifyBabelAST(value$.node),
            }
    }
  }

  return prop
}
