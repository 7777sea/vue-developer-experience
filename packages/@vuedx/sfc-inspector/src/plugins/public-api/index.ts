import { InspectorPlugin } from '../../interfaces'
import { inspectEvents } from './events'
import { inspectProps } from './props'

export const ComponentPublicAPIInspectorPlugin: InspectorPlugin = (
  descriptor,
  context
) => {
  if (!descriptor.script) return

  inspectProps(descriptor.script, context)
  inspectEvents(descriptor.script, context)
}
