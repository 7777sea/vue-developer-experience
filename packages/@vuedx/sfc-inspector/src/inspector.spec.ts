import Path from 'path'
import FS from 'fs'
import { baseInspect } from './inspector'
import { InspectorOptions } from './interfaces'
import { ComponentPublicAPIInspectorPlugin } from './plugins/public-api'

describe('@vuedx/sfc-inspector', () => {
  const context = Path.resolve(__dirname, '__fixtures__')
  const files = FS.readdirSync(context)
  const options: InspectorOptions = {
    onError(error) {
      throw error
    },
    async load(id) {
      return {
        filename: id,
        content: await FS.promises.readFile(id, { encoding: 'utf8' }),
      }
    },
    async resolve(imported, importee) {
      return Path.resolve(Path.dirname(imported), imported)
    },
    plugins: [ComponentPublicAPIInspectorPlugin],
  }

  async function doInspect(fileName: string) {
    const absoluteFilePath = Path.resolve(context, fileName)
    const source = await options.load(absoluteFilePath)
    const { info } = await baseInspect(source, options)

    return { filename: fileName, source: source.content, info }
  }

  describe(`inspect()`, () => {
    test(`props: array`, async () => {
      const { info } = await doInspect('props-array.vue')

      expect(info.props).toHaveLength(2)
    })

    test(`props: object`, async () => {
      const { info } = await doInspect('props-object.vue')

      expect(info.props).toHaveLength(4)
    })

    files.forEach(file =>
      test(`snapshot(${file})`, async () => {
        const result = await doInspect(file)

        expect(result).toMatchSnapshot()
      })
    )
  })
})

expect.addSnapshotSerializer({
  test() {
    return true
  },
  serialize({ source, info }) {
    return `${source}\n<api-doc lang="json">\n${JSON.stringify(
      info,
      null,
      2
    )}\n</api-doc>`
  },
})
