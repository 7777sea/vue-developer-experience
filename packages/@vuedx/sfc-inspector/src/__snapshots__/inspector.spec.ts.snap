// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@vuedx/sfc-inspector inspect() snapshot(composition.vue) 1`] = `
<script>
import { defineComponent } from 'vue'

export default defineComponent({
  props: ['foo'],
})
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "any"
      },
      "meta": {}
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(derived-composition.vue) 1`] = `
<script>
import { createComponent } from 'random-vue'

export default createComponent({
  props: ['foo'],
})
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "any"
      },
      "meta": {}
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(options-identifier.vue) 1`] = `
<script>
const component = {
  props: ['foo'],
}

export default component
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "any"
      },
      "meta": {}
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(options-object.vue) 1`] = `
<script>
export default {
  props: ['foo'],
}
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "any"
      },
      "meta": {}
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(props-array.vue) 1`] = `
<script>
export default {
  props: ['foo', 'bar'],
}
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "any"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "any"
      },
      "meta": {}
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(props-empty.vue) 1`] = `
<script>
export default {
  props: {},
}
</script>

<api-doc lang="json">
{
  "props": [],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(props-none.vue) 1`] = `
<script>
export default {}
</script>

<api-doc lang="json">
{
  "props": [],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(props-object.vue) 1`] = `
<script>
export default {
  props: {
    foo: String,
    bar: [String, Number],
    baz: {
      type: String,
    },
    bam: {
      type: [String, Number, Boolean],
      required: true,
    },
  },
}
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "string|number",
        "static": "string|number"
      },
      "meta": {}
    },
    {
      "name": "baz",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bam",
      "description": "",
      "required": true,
      "default": null,
      "type": {
        "runtime": "string|number|boolean",
        "static": "string|number|boolean"
      },
      "meta": {}
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(typescript-composition-context-events.vue) 1`] = `
<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    foo: String,
    bar: Number,
  },
  setup(props, context) {
    function doSomething() {
      context.emit('event1', props.foo)
    }

    return () => doSomething()
  },
})
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "number",
        "runtime": "number"
      },
      "meta": {}
    }
  ],
  "events": [
    {
      "name": "event1",
      "description": "",
      "type": {
        "static": "any"
      }
    }
  ]
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(typescript-composition-context-rest-events.vue) 1`] = `
<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    foo: String,
    bar: Number,
  },
  setup(props, { slots, ...context }) {
    function doSomething() {
      context.emit('event1', props.foo)
    }

    return () => doSomething()
  },
})
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "number",
        "runtime": "number"
      },
      "meta": {}
    }
  ],
  "events": [
    {
      "name": "event1",
      "description": "",
      "type": {
        "static": "any"
      }
    }
  ]
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(typescript-composition-events.vue) 1`] = `
<script lang="ts">
import { defineComponent } from 'vue'

/**
 * @emits {string} event2 Emitted when something happens.
 */
export default /** @emits {number} event3 */ defineComponent({
  props: {
    foo: String,
    bar: Number,
  },
  setup(props, { emit }) {
    function doSomething() {
      emit('event1', props.foo)
    }

    return () => doSomething()
  },
})
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "number",
        "runtime": "number"
      },
      "meta": {}
    }
  ],
  "events": [
    {
      "name": "event2",
      "description": "Emitted when something happens.",
      "type": {
        "static": "string"
      }
    },
    {
      "name": "event3",
      "description": "",
      "type": {
        "static": "number"
      }
    },
    {
      "name": "event1",
      "description": "",
      "type": {
        "static": "any"
      }
    }
  ]
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(typescript-object-events.vue) 1`] = `
<script lang="ts">
/**
 * @emits {string} event2 Emitted when something happens.
 */
export default /** @emits {number} event3 */ {
  props: {
    foo: String,
    bar: Number,
  },
  methods: {
    created() {
      this.$emit('event1', this.foo)
    },
  },
  created() {
    /** Emitted when created. */
    this.$emit('event4', this.foo as { bar: string })
  },
}
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "number",
        "runtime": "number"
      },
      "meta": {}
    }
  ],
  "events": [
    {
      "name": "event2",
      "description": "Emitted when something happens.",
      "type": {
        "static": "string"
      }
    },
    {
      "name": "event3",
      "description": "",
      "type": {
        "static": "number"
      }
    },
    {
      "name": "event1",
      "description": "",
      "type": {
        "static": "any"
      }
    },
    {
      "name": "event4",
      "description": "Emitted when created.",
      "type": {
        "static": "{bar:string;}"
      }
    }
  ]
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(typescript-props-object.vue) 1`] = `
<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    foo: String,
    bar: [String, Number],
    /** Merge comments */
    /**
     * @type {{foo: string}}
     * @param {{foo: string}} foo hello world
     */
    baz: {
      type: Object,
      default: () => ({ foo: 'just do it!' }),
    },
    /**
     * It's a prop.
     * @deprecated
     */
    bam: {
      type: [String, Number, Boolean],
      required: true,
      default(props) {
        return props.foo + 5
      },
    },
    /**
     * Nope
     * @private
     */
    bak: {
      type: [String, null],
      default: 'just do it!',
    },
  },
})
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "runtime": "string|number",
        "static": "string|number"
      },
      "meta": {}
    },
    {
      "name": "baz",
      "description": "Merge comments",
      "required": false,
      "default": {
        "factory": "()=>{foo:'just do it!'}"
      },
      "type": {
        "static": "{foo:string}",
        "runtime": "object"
      },
      "meta": {}
    },
    {
      "name": "bam",
      "description": "It's a prop.",
      "required": true,
      "default": {
        "factory": "function(props){return props.foo+5;}"
      },
      "type": {
        "runtime": "string|number|boolean",
        "static": "string|number|boolean"
      },
      "meta": {
        "deprecated": true
      }
    },
    {
      "name": "bak",
      "description": "Nope",
      "required": false,
      "default": {
        "value": "'just do it!'"
      },
      "type": {
        "runtime": "string|null",
        "static": "string|null"
      },
      "meta": {
        "private": true
      }
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(typescript-props-type-inline.vue) 1`] = `
<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent<{
  foo: string
  bar: { foo: string; bar: number }
}>({
  props: ['foo', 'bar'],
})
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": true,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": true,
      "default": null,
      "type": {
        "runtime": "any",
        "static": "{foo:string;bar:number;}"
      },
      "meta": {}
    }
  ],
  "events": []
}
</api-doc>
`;

exports[`@vuedx/sfc-inspector inspect() snapshot(typescript-render-slots.vue) 1`] = `
<script lang="ts">
/**
 * @emits @param {string} event2 Emitted when something happens.
 * @slot @param {{ data: string }} default The default slot.
 */
export default /** @emits {number} event3 */ {
  props: {
    foo: String,
    bar: Number,
  },
  methods: {
    created() {
      this.$emit('event1', this.foo)
    },
  },
  created() {
    /** Emitted when created. */
    this.$emit('event4', this.foo as { bar: string })
  },
}
</script>

<api-doc lang="json">
{
  "props": [
    {
      "name": "foo",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "string",
        "runtime": "string"
      },
      "meta": {}
    },
    {
      "name": "bar",
      "description": "",
      "required": false,
      "default": null,
      "type": {
        "static": "number",
        "runtime": "number"
      },
      "meta": {}
    }
  ],
  "events": [
    {
      "name": "event2",
      "description": "Emitted when something happens.",
      "type": {
        "static": "string"
      }
    },
    {
      "name": "event3",
      "description": "",
      "type": {
        "static": "number"
      }
    },
    {
      "name": "event1",
      "description": "",
      "type": {
        "static": "any"
      }
    },
    {
      "name": "event4",
      "description": "Emitted when created.",
      "type": {
        "static": "{bar:string;}"
      }
    }
  ]
}
</api-doc>
`;
